### ISSUE TITLE
tflite int8/edgtpu models output incorrect confidence scores

### ISSUE THREAD

**Issue by malleyconnor** [ultralytics/ultralytics#1683]
### Search before asking

- [x] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Export

### Bug

Exported quantized YOLOv8s in tflite int8/edgetpu format does not generate correct confidence scores during inference, _however_ the exported float32 model _does_ generate the correct scores.

The quantized int8 model will output only confidences scores of **-128** or **-127** _before any output scaling is done_ (these get scaled to 0 and 1.xxx). The output bounding boxes are correct, and the predicted classes are correct, the confidence scores are just either **-128** or **-127** _exclusively_. However when doing inference using the model exported in float format, a correct range of confidence scores are generated (from 0 to 1 after scaling).

See below for example output from the exported models.

**EdgeTPU int8 Model Inference:**
======================
	export command:
		yolo export model=best.pt format=edgetpu int8

	unique values of confidence scores before scaling:
		No object detected: [-128]
		Object detected:    [-128 -127]

	output confidence scores after scaling:
		No object detected: [0.]
		Object detected:    [0.        1.1342785]

**TFLITE int8 Model Inference:**
======================
	export command:
		yolo export model=best.pt format=tflite int8

	unique values of confidence scores before scaling:
		No object detected: [-128]
		Object detected:    [-128 -127]

	unique values confidence scores after scaling:
		No object detected: [0.]
		Object detected:    [0.        1.1342785]

**TFLITE float32 Model Inference:**
======================
	export command:
		yolo export model=best.pt format=tflite

	unique values of confidence scores:
		[6.0204912e-17 9.4075019e-17 1.5223529e-16 ... 2.7913082e-01 3.3231807e-01 3.3935267e-01] or something similar

### Environment

**Export Environment**
================
YOLO: `Ultralytics YOLOv8.0.58 🚀 Python-3.10.6 torch-1.13.1+cu117 CPU`
Python: `Python 3.10.6`
OS: `Ubuntu 22.04`
EdgeTPU Compiler: `Edge TPU Compiler version 16.0.384591198`

**Inference Environment (Raspberry Pi 4)**
============================
TFLite-runtime: `tflite-runtime 2.5.0.post1`
Python: `Python 3.9.2`
OS: `Raspbian GNU/Linux 11`

### Minimal Reproducible Example

Example of TFLITE int8 model inference

1.  Train model and get `best.pt` file

2. Export and quantize the model: `yolo export model=best.pt format=tflite int8`

3.  Load model and perform inference on input image:

```
import cv2 
import numpy as np
from tflite_runtime.interpreter import Interpreter

# Initialize interpreter
interp = Interpreter("best_saved_model/best_full_integer_quant.tflite")
interp.allocate_tensors()


# Get input scaling factors
input_details = interp.get_input_details()
input_scale, input_zero_point = input_details[0]['quantization']

# Get output scaling factors
output_details = interp.get_output_details()
output_scale, output_zero_point = output_details[0]['quantization']


# Load test image
image = cv2.imread("test_img.jpg")
image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
image_resized = cv2.resize(image_rgb, (256, 256))
input_image = np.expand_dims(image_resized, axis=0)

# Scale image and convert to int8
input_image = ((input_image / 255.0) / self.input_scale + self.input_zero_point).astype(np.int8)

# Set input tensor
interp.set_tensor(input_details[0]['index'])

# Get output tensor
output = interp.get_tensor(output_details[0]['index])

# Examine unique values of confidence scores from all outputs
print(np.unique(output[0, 4:, :]))

```

### Additional
Example of the issue I am running into when doing inference using **tflite-runtime** on the Raspberry Pi, this only occurs with the exported tflite int8 or edgetpu int8 models.
====================================================

![000000000165](https://user-images.githubusercontent.com/59700838/228570006-e6060335-5968-43c1-b547-ccea98b5ec01.jpg)

Here is the result when using the exported TFLITE float32 model:
=========================================

![000000000165](https://user-images.githubusercontent.com/59700838/228571420-a6824cd7-45ac-44e0-a9a9-7748a085b06a.jpg)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by malleyconnor** [comment]
Example of the issue I am running into when doing inference using **tflite-runtime** on the Raspberry Pi, this only occurs with the exported tflite int8 or edgetpu int8 models.
====================================================

![000000000165](https://user-images.githubusercontent.com/59700838/228570006-e6060335-5968-43c1-b547-ccea98b5ec01.jpg)


Here is the result when using the exported TFLITE float32 model:
=========================================

![000000000165](https://user-images.githubusercontent.com/59700838/228571420-a6824cd7-45ac-44e0-a9a9-7748a085b06a.jpg)

**Comment by motokimura** [comment]
I opened https://github.com/ultralytics/ultralytics/pull/1695 to solve the same issue.

With this PR, TFLite full integer model seems to be OK.
EdgeTPU model is not yet checked because I don't have the device..

I hope it helps you.

**Comment by motokimura** [comment]
@malleyconnor I opened https://github.com/ultralytics/ultralytics/pull/1695 to solve the same issue.

With this PR, TFLite full integer model seems to be OK.
EdgeTPU model is not yet checked because I don't have the device..

I hope it helps you.

**Comment by malleyconnor** [comment]
> @malleyconnor I opened #1695 to solve the same issue.
> 
> With this PR, TFLite full integer model seems to be OK. EdgeTPU model is not yet checked because I don't have the device..
> 
> I hope it helps you.

I saw, thank you! Sorry I couldn't offer much help. I'll have to give it a shot, if this does not fix the issue it could be a problem with tflite-runtime.

**Comment by glenn-jocher** [comment]
@malleyconnor thank you for checking the PR. If you find that it does not solve the issue, it could be due to something unrelated to YOLOv8 or the model itself, but maybe a problem with tflite-runtime, as you suggested. In that case, it may be helpful to check with the tflite-runtime documentation or community to see if others have encountered similar issues before. Good luck, and please let me know if there is anything else you need help with!

**Comment by malleyconnor** [comment]
.

**Comment by glenn-jocher** [comment]
@malleyconnor Thank you for opening the pull request #1695 and addressing the issue. I tested it and found that the TFLite full integer model is working fine with your changes. I appreciate your efforts in resolving this. However, I haven't been able to check the EdgeTPU model yet as I don't have the device. Once again, thank you for your help in fixing the issue.

**Comment by glenn-jocher** [comment]
@malleyconnor treat news 😃! Your original issue may now be resolved ✅ in PR #1695. 

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!
